faidx.c

```diff
@@ -124,69 +152,85 @@ static int write_line(faidx_t *faid, FILE *file, const char *line, const char *n
     for (i = 0; i < seq_sz; i += length)
     {
         hts_pos_t len = i + length < seq_sz ? length : seq_sz - i;
-        if (fwrite(line + i, 1, len, file) < len ||
-            fputc('\n', file) == EOF) {
-            print_error_errno("faidx", "failed to write output");
-            return EXIT_FAILURE;
-        }
+        if (_wrappedwrite(out, line + i, len) < len ||
+                _wrappedwrite(out, "\n", 1) < 1) {     //writing \n alone
+                print_error_errno("faidx", "failed to write output");
+                return EXIT_FAILURE;
+            }
     }
 
     return EXIT_SUCCESS;
 }
 
 
-static int write_output(faidx_t *faid, FILE *file, const char *name, const int ignore,
+static int write_output(faidx_t *faid, output *out, const char *name, const int ignore,
                         const hts_pos_t length, const int rev,
                         const char *pos_strand_name, const char *neg_strand_name,
                         enum fai_format_options format) {
-    hts_pos_t seq_len, wrap_len = length;
+    hts_pos_t seq_len, wrap_len = length, namelen = 0;
+    char *seq =  NULL, *qual = NULL;
+    int ret = EXIT_FAILURE;
+
     if (wrap_len < 0)
         wrap_len = fai_line_length(faid, name);
     if (wrap_len <= 0)
         wrap_len = HTS_POS_MAX;
-    char *seq = fai_fetch64(faid, name, &seq_len);
 
-    if (format == FAI_FASTA) {
-        fprintf(file, ">%s%s\n", name, rev ? neg_strand_name : pos_strand_name);
-    } else {
-        fprintf(file, "@%s%s\n", name, rev ? neg_strand_name : pos_strand_name);
+    //make buffer to format the name
+    namelen = 3 + strlen(name) + strlen(rev ? neg_strand_name : pos_strand_name);   // >/@ + namelen + strand len + \n + null
+    if (!out->buffer || out->buff_len < namelen) {
+        out->buffer = realloc(out->buffer, namelen);
+        out->buff_len = namelen;
+    }
+    if (!out->buffer) {
+        fprintf(stderr,"[faidx] Couldn't create buffer\n");
+        goto exit;
     }
```

Comment:
I'm wondering if we should just create a `bgzf_putline` to be analogous the the `bgzf_getline` function so we can use kstring instead of handling our own dynamic reallocation buffer.

Or if not, then kstring here directly instead.  That would mean the existing fprintfs just become ksprintfs and the code change is relatively lightweight.

Reply:
changed to use kstring.